core/src/main/scala/org/http4s/util/StreamApp.scala

```diff
@@ -4,56 +4,88 @@ import cats.effect._
+            .takeWhile(_ == false)
+            .run
         hook.unsafeRunSync()
       }
       ()
     }
 
   /** Exposed for testing, so we can check exit values before the dramatic sys.exit */
-  private[util] def doMain(args: List[String]): IO[Int] = {
+  private[util] def doMain(args: List[String]): IO[ExitCode] = {
     implicit val ec: ExecutionContext = execution.direct
-    async.ref[IO, Int].flatMap { exitCode =>
-      async.signalOf[IO, Boolean](false).flatMap { halted =>
-        async
-          .signalOf[F, Boolean](false)
-          .flatMap { requestShutdown =>
-            addShutdownHook(requestShutdown, halted) >>
-              stream(args, requestShutdown.set(true))
-                .interruptWhen(requestShutdown)
-                .run
-          }
-          .runAsync {
-            case Left(t) =>
-              IO(logger.error(t)("Error running stream")) >>
-                halted.set(true) >>
-                exitCode.setSyncPure(-1)
-            case Right(_) =>
-              halted.set(true) >>
-                exitCode.setSyncPure(0)
-          } >>
-          exitCode.get
-      }
-    }
+    for {
+      exitCodeRef <- async.ref[IO, ExitCode]
+      halted <- async.signalOf[IO, Boolean](false)
+      exitCode <- runStream(args, exitCodeRef, halted)
+    } yield exitCode
+  }
+
+  /**
+    * Runs the application stream to an ExitCode.
+    *
+    * @param args The command line arguments
+    * @param exitCodeRef A ref that will be set to the exit code from the stream
+    * @param halted A signal that is set when the application stream is done
+    * @param ec Implicit EC to run the application stream
+    * @return An IO that will produce an ExitCode
+    */
+  private[util] def runStream(
+      args: List[String],
+      exitCodeRef: Ref[IO, ExitCode],
+      halted: Signal[IO, Boolean]
+  )(implicit ec: ExecutionContext): IO[ExitCode] = {
+    val runStreamLast: F[Option[ExitCode]] =
+      for {
+        requestShutdown <- async.signalOf[F, Boolean](false)
+        _ <- addShutdownHook(requestShutdown, halted)
+        exitCode <- stream(args, requestShutdown.set(true))
+          .interruptWhen(requestShutdown)
+          .take(1)
+          .runLast
+      } yield exitCode
+    runStreamLast.runAsync {
+      case Left(t) =>
+        IO(logger.error(t)("Error running stream")) >>
+          halted.set(true) >>
+          exitCodeRef.setSyncPure(ExitCode.error)
+      case Right(exitCode) =>
+        halted.set(true) >>
+          exitCodeRef.setSyncPure(exitCode.getOrElse(ExitCode.success))
+    } >>
+      exitCodeRef.get
   }
 
   def main(args: Array[String]): Unit =
-    sys.exit(doMain(args.toList).unsafeRunSync)
+    sys.exit(doMain(args.toList).unsafeRunSync.code)
+}
+
+object StreamApp {
+  sealed abstract case class ExitCode(code: Int)
+
+  object ExitCode {
+    def fromInt(code: Int): ExitCode = new ExitCode(code) {}
+    val success: ExitCode = fromInt(0)
+    val error: ExitCode = fromInt(-1)
```

Comment:
I think of 1 as the generic error code, not -1 (aka 255).  Though I've seen both.  I suppose it would be nice to distinguish apps that died by exception vs. apps that were asked to do something they can't.  Sort of like failure vs. error in a test framework.  But I can't point to a standard for that. 

2 is traditional for command line usage failure and might be worth a constant.

Reply:
I found this: http://tldp.org/LDP/abs/html/exitcodes.html

I guess most of them doesn't really make sense. I will switch error to 1 instead.